### DISCUSSION TITLE
(remotely-save author here) some thoughts on this project

### DISCUSSION THREAD

**Discussion by fyears** [sboesen/remotely-sync#76] [General]
Hi @sboesen !

I am @fyears the original author of remotely-save. Though I am still using Obsidian and `remotely-save` for personal projects, I haven't paid attention to the development for a long while... Today I come across your project and I would like to share some thoughts.

1. First of all, great work for maintaining the forked plugin! That's why I like participating in the open source and Obsidian community. 👍 
2. The encryption method was changed (original idea of this project?). That's ok. But I notice that you removed the some testing suite and the `openssl` part of document. 
    * I **deliberately** kept the algorithm reproducible in openssl, because encryption is hard, and the plugin may corrupt. It's important to generate the encryption file compatible with well-known tools instead of a home-made solution. 
    * More importantly, **the `iv` should be different because [the `salt` is different](https://github.com/remotely-save/remotely-save/blob/b238988012f63f851c56496263cf469237e78a29/src/encrypt.ts#L50)**. So in my opinion that’s not a security problem.
    * Furthermore, it might be better to provide a progressive upgrade of users' files instead of requiring re-uploading everything.
    * I am also thinking of providing better encryption and might develop some improvement in `remotely-save`.
3. I **deliberately** avoid the obvious name [`remotely-sync`](https://github.com/obsidianmd/obsidian-releases/pull/2737) in the very beginning because it's very easy to be confused with official Obsidian Sync service. To be honest I don't want to make Obsidian team unhappy and I would like to avoid any legal issues... 
4. Some of notices and backward capabilities were removed in your fork. But why?
5. You embed some hard-code environment variables in your code, I would suggest removing them and feed them by GitHub environment variables.
6. It seems that some extra metadata was uploaded to s3 (for modification time?). But will this solution applicable to other remote like webdav? IMO it's important to gain similar behaviors across different remotes.
7. It’s strongly welcome to open one or many PR, and I would like to merge the changes if applicable!!!

**Comment by kadisonm** [comment]
Hi @fyears nice to meet you.

Just commenting with a question about the PR. After merging would there still be a way to continue improving upon remotely save. One thing I definitely liked was the consistent updates to this plugin and it would be great if that continued with remotely save. But I also understand if you can't actively maintain the project that frequently.

**Comment by sboesen** [comment]
@fyears thank you for the comment! I will respond point by point later today, just wanted to say I really appreciate you building Remotely Save and for the feedback in this comment.

**Comment by sboesen** [comment]
1. Thanks! Very grateful for you writing the original plugin and maintaining it for the period you did.
2. 
**Reproducibility:**
> I deliberately kept the algorithm reproducible in openssl, because encryption is hard, and the plugin may corrupt. It's important to generate the encryption file compatible with well-known tools instead of a home-made solution.

Thankfully AES-GCM is pretty standard and not home-made, and [included in openssl](https://security.stackexchange.com/questions/30344/why-does-openssl-not-include-aes-256-gcm) but I do need to re-add documentation and testing code. Definitely planned (#10)!

**IV:**
It looks like I misread the encryption routine and made the incorrect assumption the saltHex parameter was used - sorry about that & appreciate the callout, definitely a mistake doing code review! Meant to test this further in a writeup and got sidetracked updating this plugin.

I see where you added a test to ensure encryption is not random, and that makes sense to me (https://github.com/remotely-save/remotely-save/commit/c689235ff0909e264c6d378bd636aad20df74f37).

I think the salt should be 16 bytes (it is [8 bytes](https://github.com/remotely-save/remotely-save/blob/b238988012f63f851c56496263cf469237e78a29/src/encrypt.ts#L50)) per specification of [Pbkdf2Params](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), but that does make it random even at 8 bytes (and 2^64 is pretty low chance of repeating) but I don't have a full intuition if that is safe. Instead of deriving IV from output of key derivation with random salt a suggestion - why not generate an IV similarly to the salt:

`  let salt = window.crypto.getRandomValues(new Uint8Array(16));`
`  let iv = window.crypto.getRandomValues(new Uint8Array(12));` // Or whatever length you need, in GCM it is 12 bytes

Will update Remotely Sync's readme to reflect my latest understanding of the encryption changes.

**Backwards compatibility:**
> Furthermore, it might be better to provide a progressive upgrade of users' files instead of requiring re-uploading everything.

Totally understand, I don't want to maintain multiple copies of encryption routines due to the testing demands and change to the algorithm but understand your choice to do so. I think making the suggested changes will improve security of Remotely Save but will require some additional checks for older salt/iv values, etc.

**Encryption improvements:** I do think GCM is an improvement, but the IV -> not derived from key change doesn't seem like as high of a priority given the salt is random (though could be more bits per spec as I mentioned).

3. Definitely agreed, I'm more than happy to change the name again [if they have any issues](https://github.com/obsidianmd/obsidian-releases/pull/2737) with the name. Thankfully I checked and "Sync" is used in other approved plugin names, so I think it should be OK.

4. Backwards compatibility was removed to simplify the code and make it easier to update - originally this plugin was just meant for me, so I didn't need backwards compatibility! I wouldn't recommend merging these changes into Remotely Save due to the large number of users.

5. I'm not too concerned about this as these are just client IDs and I don't think they will change between say a test environment and production, also not using GitHub to build releases currently (doing it by hand D:)
6. This is [planned](https://github.com/sboesen/remotely-sync/issues/70), some providers don't support adding metadata which I realized part-way through. I plan to add this to the metadata file with an entry for each pair of (filename, mtime). This will help add other providers & reduce provider-specific sync complexity. 
7. More than happy to open PRs - I think it's worth discussing what you're interested in taking for Remotely Save as I think we have slightly different goals (backwards compatibility not being one for me, but understand with the large user base of Remotely Save it will be a challenge). Several of the changes have existing PRs for Remotely Save (or are implemented in other forks) so happy to point you in that direction, too!

It may take a bit of time to open isolated PRs because the plugin has diverged a bit and several new features which interact with each other have been added, like syncing trash folder/deleting trash folder & attempts to refactor the sync routine to make it easier to do a partial sync. Other features are WIP which works fine for a small fork but probably not worth merging until they are complete in the main repo.

@kadisonm 
> After merging would there still be a way to continue improving upon remotely save. One thing I definitely liked was the consistent updates to this plugin and it would be great if that continued with remotely save. But I also understand if you can't actively maintain the project that frequently.

100% - I'm happy to contribute changes but I suspect the philosophy of keeping things working for existing users has a different result than our plugin's "move fast and break things" - We don't have regular beta releases and I try to fix things within a day or two if they break, but that obviously won't scale. I plan to keep this fork maintained but that doesn't stop us (or @fyears) from merging some changes upstream!

**Comment by sboesen** [comment]
Updated readme!

https://github.com/sboesen/remotely-sync/commit/9bd910b25784096bf1887f9bbf148295fd32e76f

**Comment by kadisonm** [comment]
> I plan to keep this fork maintained but that doesn't stop us (or @fyears) from merging some changes upstream!

Ah okay. I like the idea of Remotely Save being a stable version with some of the more polished features in this plugin being merged into it when they're ready.

**Comment by terminar** [comment]
Regarding 3.: i think there only some small things where @kepano / the Obsidian team would not be happy with:
____________
What we're not okay with:

    Using Obsidian trademarks and code in a way that could confuse users into thinking it's an official solution
    Leaving security vulnerabilities open
    Reverse engineering for the purpose of copying our code to make commercial solutions, see the [Restrictions](https://obsidian.md/terms#restrictions) section of our terms

____________

What we're okay with:

    Reverse engineering for fun, as a technical challenge, or to help build a plugin
    Making plugins/services that let users sync/publish for free (there are many of them already)
    Submitting those plugins to the official directory, as long as they meet our [developer policies](https://docs.obsidian.md/Developer+policies)

____________

https://github.com/acheong08/obi-sync/issues/19#issuecomment-1702926188

They don't seem to have a problem with using "sync" in the name of a plugin.
